doc/i18n.md

```diff
@@ -126,3 +163,231 @@ plugin which automatically updates all PO files to match the POT file. That
+
+#### Singular Form
+
+This is the most common way:
+
+```js
+const title=_("Storage");
+```
+
+For using translated text in React components you need to use curly braces
+to switch to plain Javascript in the attributes and the content:
+
+```js
+<Section title={_("Storage")}></Section>
+```
+
+or
+
+```js
+<Button>{_("Select")}</Button>
+```
+
+For translating long texts you might use multiline string literals:
+
+```js
+const description = _("Select the device for installing the system. All the \
+file systems will be created on the selected device.");
+```
+
+If you need to insert some values into the text you should use the
+[formatting function](#formatting-texts). It is recommended to use
+[translator comments](#translators-comments) for short or ambiguous texts.
+
+#### Plural Form
+
+If the translated text contains a number or quantity it is good to use a plural
+form so the translated texts looks naturally.
+
+```js
+sprintf(
+  // the first argument is a singular form used when errors.length === 1,
+  // the second argument is used in all other cases
+  n_("%d error found", "%d errors found", errors.length),
+  errors.length
+)
+```
+
+Although the English language has only a single plural form the translators can
+use more of them. The translation definition also defines a function that
+computes which plural form to use for a specific number. See more details about
+the [plural forms](
+https://www.gnu.org/software/gettext/manual/html_node/Plural-forms.html) in the
+GNU gettext documentation.
+
+#### Translating Constants
+
+The top-level constants are evaluated too early, at that time the translations
+are not available yet and the language is not set.
+
+The constant value must be marked with `N_()` function which does nothing
+and then later translated with the usual `_()` function.
+
+```js
+const LABELS = Object.freeze({
+  auto: N_("Auto"),
+  fixed: N_("Fixed"),
+  range: N_("Range")
+});
+
+export default function Foo() {
+  ...
+  <label>{_(LABELS[value])}</label>
+```
+
+#### Formatting Texts
+
+For formatting complex texts use the C-like `sprintf()` function.
+
+```js
+import { sprintf } from "sprintf-js";
+
+sprintf(_("User %s will be created"), user)
+```
+
+See the [sprintf-js](https://www.npmjs.com/package/sprintf-js) documentation
+for more details about the formatting specifiers.
+
+It is recommended to avoid building texts from several parts translated
+separately. This is error prone because the translators will not see the
+complete final text and might translate some parts inaccurately.
+
+```js
+// do NOT use this! it is difficult to translate the parts correctly
+// so they build a nice text after merging
+return <div>{_("User ")}<b>{user}</b>{_(" will be created")}</div>
+
+// rather translate a complete text and then split it into parts
+// TRANSLATORS: %s will be replaced by the user name
+const [msg1, msg2] = _("User %s will be created").split("%s");
+...
+return <div>{msg1}<b>{user}</b>{msg2}</div>
```

Comment:
That does not work, that `bold_user` constant is not a string, it is an object. This would be rendered as:
```
User [object Object] will be created
```

To avoid code injection React by default escapes all inserted texts so you cannot use HTML tags in translations:

```
{sprintf(_("User <b>%s</b> will be created"), "USER")}
```

is rendered as `User &lt;b&gt;USER&lt;/b&gt; will be created` HTML so the user will see:

```
User <b>USER</b> will be created
```

There is a way to inject unescaped text but it is insecure and not recommended: https://legacy.reactjs.org/docs/dom-elements.html#dangerouslysetinnerhtml

Reply:
Ah so what I was missing is: in React you cannot make HTML from strings, only from React objects